docs/source/notes/amp_examples.rst

```diff
@@ -0,0 +1,158 @@
+
+
+Working with Unscaled Gradients
+-------------------------------
+
+The gradients resulting from ``scaler.scale(loss).backward()`` are scaled.  ``scaler.unscale`` can be
+used to unscale an optimizer's owned gradients in-place prior to ``scaler.step``.
+
+Gradient clipping
+"""""""""""""""""
+
+Gradient clipping involves manipulating a set of gradients such that their global norm
+(see :func:`torch.nn.utils.clip_grad_norm_`) or maximum magnitude (see :func:`torch.nn.utils.clip_grad_value_`)
+is :math:`<=` some user-imposed threshold.  Calling ``scaler.unscale(optimizer)`` before clipping enables you
+to clip unscaled gradients as usual::
+
+    scaler = AmpScaler()
+    ...
+    for input, target in data:
+        optimizer.zero_grad()
+        output = model(input)
+        loss = loss_fn(output, target)
+        scaler.scale(loss).backward()
+
+        # Unscale the gradients owned by optimizer in-place
+        scaler.unscale(optimizer)
+
+        # Since the optimizer's owned gradients are unscaled, clip as usual:
+        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm)
+
+        scaler.step(optimizer)
+        scaler.update()
+
+``scaler`` records that ``scaler.unscale(optimizer)`` has been called for this optimizer
+this iteration, so ``scaler.step(optimizer)`` knows not to redundantly unscale gradients before
+calling ``optimizer.step()``.
+
+.. warning::
+    If you (optionally) choose to unscale gradients prior to stepping, ``scaler.unscale(optimizer)``
+    should only be invoked once per optimizer per step, and only after all gradients for that optimizer's
+    owned parameters have been accumulated.
+
+Working with Scaled Gradients
+-----------------------------
+
+For some operations, you may need to work with scaled gradients in a setting where
+``scaler.unscale`` is unsuitable.
+
+Gradient penalty
+""""""""""""""""
+
+A gradient penalty loss term requires manipulating scaled gradients.
+The gradient penalty example below demonstrates
+
+* creating scaled out-of-place gradients with :func:`torch.autograd.grad`, and
+* correct interaction of gradient scaling with double-backward.
+
+Here's how that looks for a simple L2 penalty::
+
+    scaler = AmpScaler()
+    ...
+    for input, target in data:
+        optimizer.zero_grad()
+        output = model(input)
+        loss = loss_fn(output, target)
+
+        # We should scale outputs for the out-of-place backward pass
+        grad_params = torch.autograd.grad(scaler.scale(loss), model.parameters(), create_graph=True)
+
+        # In general, the penalty term may depend nonlinearly on grad_params, so to be safe,
+        # manually unscale them before computing the penalty.  The unscale should be out-of-place
+        # and autograd-exposed.  For these reasons, and because grad_params are not owned by any optimizer,
+        # calling scaler.unscale(optimizer) here is unsuitable, and we use ordinary ops instead:
+        inv_scale = 1./scaler.get_scale()
```

Comment:
Greg: why now always expose unscale to autograd?

Reply:
Greg: unscale(reentract = true)?

Reply:
Greg: maybe inplace and not inplace version of this function? E.g. unscale_ and unscale

Reply:
Unscale has a narrow purpose:  to unscale gradients owned by a particular optimizer and check for infs/nans, without moving them, after all autograd work has completed.  Here the purpose is different:  unscale gradients not owned by any optimizer, in an out-of-place and autograd-exposed way.

I think it would be more confusing to create overloads of `unscale` that serve a different purpose or purposes than to have people use existing, familiar operations for the different purpose.

I can rename `unscale` to `unscale_` if you think that's clearer.  Also, additional options for `unscale` could be added as sugar in later PRs, but for this PR, I would like to keep it focused on its usual purpose.